kernel/src/main/scala/me/rexim/morganey/syntax/LambdaParser.scala

```diff
@@ -27,38 +28,65 @@ class LambdaParser extends JavaTokenParsers with ImplicitConversions {
   def variable: Parser[LambdaVar] =
     identifier.r ^^ { LambdaVar }
 
-  def numericLiteral: Parser[LambdaTerm] =
+  def validNumberLiteral: Parser[Int] =
     numberLiteral.r ^? ({
-      case IntMatcher(x) => ChurchNumberConverter.encodeNumber(x)
+      case IntMatcher(x) => x
     }, { (rawInt) => s"`$rawInt' is too big"})
 
-  def characterLiteral: Parser[LambdaTerm] = (
+  def numericLiteral: Parser[LambdaTerm] =
+    validNumberLiteral ^^ ChurchNumberConverter.encodeNumber
+
+  def validCharacterLiteral: Parser[Char] = (
     escapedCharLiteral.r ^^ { s =>
-      ChurchNumberConverter.encodeNumber(escapeSequences(s charAt 2))
-    }
-    | symbolCharLiteral.r ^^ { s =>
-      ChurchNumberConverter.encodeNumber(s charAt 1)
+      escapeSequences(s charAt 2)
     }
+    | symbolCharLiteral.r ^^ { _ charAt 1 }
   )
 
+  def characterLiteral: Parser[LambdaTerm] =
+    validCharacterLiteral ^^ { c => ChurchNumberConverter.encodeNumber(c.toInt) }
+
   def stringLiteralTerm: Parser[LambdaTerm] =
     stringLiteral ^^ { s =>
       ChurchPairConverter.encodeString(unquoteString(s))
     }
 
+  private def range[T](p: => Parser[T])(implicit ev: Integral[T]): Parser[LambdaTerm] = {
+    val q = p
+    val parser = q ~ opt(comma ~> q) ~ (rangeOperator ~> q)
+    parser ^^ { case start ~ next ~ exit =>
+      val step  = next.map(ev.minus(_, start)).getOrElse(ev.one)
+      val range = NumericRange.inclusive(start, exit, step).toList
+      val nums  = range map ev.toInt map ChurchNumberConverter.encodeNumber
+      ChurchPairConverter.encodeList(nums)
+    }
+  }
```

Comment:
And if only the target type should be specified `coerceTo[Char]` you could write something like:
```scala
def coerceTo[B: Unliftable] = new MkCoerce[B]

class MkCoerce[B](implicit unliftb: Unliftable[B]) {
  def apply[A](a: A)(implicit lifta: Liftable[A]): Option[B] = termCoerce[A, B](a)
}
```
The scala compiler will infer the type of `A` while a concrete type for `B` can be specified manually.